Introducing Android Snapshot Plugin (#78)
xmartlabs/blog: _posts/2019-03-21-Introducing-Android-Snapshot-Publisher.md

---
layout: post
title: Introducing Android Snapshot Plugin
date: 2019-03-20 12:00:00
author: Matías Irland
categories: Android, Gradle Plugin, Snapshot, Google Play, Fabric Beta
author_id: mirland

---

We're happy to announce the release of our first [open source **Gradle Android Plugin**](https://github.com/xmartlabs/android-snapshot-publisher), a plugin to **create Android Snapshot versions** in the simplest way that we know!

Here at [Xmartlabs](https://www.xmartlabs.com/) we love to create new product, add new features, improve our client apps and fix the existing issues.

> On the other side are our clients, who want to get a try our work ASAP.

Author: ```suggestion
Our awesome clients are oftentimes so exited about the work we do that they want to try it ASAP.
```

> Our awesome clients are oftentimes so exited about the work we do that they want to try it ASAP.

Reviewer: ```suggestion
Our awesome clients are often so exited about the work we do that they want to try it ASAP.
```

Reviewer 2: `oftentimes` means `many times`, I think replacing it with just `often` could be not 100% right

Reviewer: OK. I had never seen that word before

Furthermore, our QA department or some other team members could also be waiting for our changes to deploy in order to test themselves.
That's why we try to prepare and deliver new *Snapshot Builds* as soon as possible.

> Can you relate to this? If you can, you probably know that preparing and distributing new builds is not too hard but can take up valuable time.

Reviewer: ```suggestion
Do you find this familiar? If you do, you probably know that preparing and distributing new builds is not too hard but can take up valuable time.
```

For example if a crash happened you'll want to know which was the specific build that caused that crash.
1. Generating good release notes (including a changelog) to check what changed when compared to previous builds.
This is essential as you must communicate what should be tested.

Based on that we decided to create our very own [**Gradle Android Snapshot Publisher plugin!**](https://github.com/xmartlabs/android-snapshot-publisher)

## Which are the main features of this plugin?

It was designed to **prepare and distribute Android Snapshot builds**.

The main features of the preparation process are:
- Update the Android Version Name to keep track of the distributed versions.
The default behavior appends the commit identifier to the Android Version name. For instance, a version name may look similar to `1.0.2-b799fbc`.
- Create and customize rich release notes based on Git's history.

After this preparation work, the plugin will proceed to compile and publish the snapshot build.
Currently the available sources to deploy the builds are:
- [Google Play](https://play.google.com/apps/publish)
- [Fabric Beta](https://docs.fabric.io/android/beta/overview.html)

> The next picture shows an example of how the plugin prepares and deploys -in Fabric Beta- a snapshot version of a typical app.

Reviewer 2: ```suggestion
The next picture shows an example of a Build published by the Plugin.
```

> ***MOCK AN IMAGE***

Reviewer: Don't forget to replace this for the image

## What is the preparation work based on?

The plugin makes good use of your project's Git repository to perform the prep work.
To get the most out of this plugin, you have to be as neat as you can when managing it.
In my team, we are following a simple but nice Git flow.

> We have two main branches: `master` which contains the production code and `develop` which contains the newest changes that have not been released to production just yet.

Reviewer: ```suggestion
We have two main branches: `master` which contains the production code and `develop` which contains the newest changes that have not been released to production yet.
```

> Then, if we want to add a new feature or fix an already existing issue, we create a new branch from `develop` and start coding!

Reviewer: ```suggestion
Then, if we want to add a new feature or fix an existing issue, we create a new branch from `develop` and start coding!
```

The next step is to create a new pull request, and wait for the reviewers to review the code we want to merge.
After we get approval, we can integrate our changes into the `develop` branch. But, how should we really integrate them? 
GitHub provides 3 predefined ways to do that, "Merge", "Rebase and Merge" or "Squash and Merge" (if you are not using GitHub you can just use Git's commands to get the same result).
Based on how the branch was created it could contain multiple commits but all of them related to only one feature or fix. 
So the main value of that branch is the feature or fix you are introducing.
For this reason, some time ago we decided to start using the "Squash and Merge" option, together with a good and descriptive message.

> The key to this flow is that if you check the Git history of your main branches, each commit has an understandable and important value.

Reviewer: I think you are introducing our Git flow in this paragraph but not concluding what the user has to do. Are these recommendations? Are you only telling them what we do? Or do they have to follow this exactly?

Author: No, the idea is that if you follow a similar flow will get a rich release notes.
@mats-claassen What about something like this:?
```suggestion
The key to this flow is that if you check the Git history of your main branches, each commit has an understandable and important value.
It was just an introduction of our Git's flow, follow it is not mandatory to use this plugin.
However, if you use a similar flow where each commit has an important value and a descriptive message your Snapshot Build** will be great! 
```

Reviewer 3: The only things that matters here is that the repository branch used to take the release note must have clear commits and ideally one commit per feature or fix. Our approach is irrelevant in terms of library usage.

## How to get started?

The first step is to [include the plugin dependency in your project and apply it in your Android module](https://github.com/xmartlabs/android-snapshot-publisher#installation), it's the typical setup of any plugin.

Then, the plugin defines a `snapshotPublisher` block where you can add the different setup alongside the Android modules.

```groovy
snapshotPublisher {
    version {
        // Version customization
    }
    releaseNotes {
        // Release notes customization
    }
    fabric {
        // Fabric Beta setup
    }
    googlePlay {
        // Google Play setup
    }
}
```

`version`: Defines the Android Version Name customizations for the delivered build.
The default value is the current version name and the short-hash commit, joined by a hyphen.

> `releaseNotes`: Define the release notes customizations.

Reviewer 2: ```suggestion
`releaseNotes`: Defines the release notes customizations.
```

`fabric`: Defines the configuration needed to deploy the artifacts in Fabric's Beta.
`googlePlay`: Defines the configuration needed to deploy the artifacts in Google Play.

> You can read the more about each block in the [project's GitHub Repository](https://github.com/xmartlabs/android-snapshot-publisher#setup), but I promise you that the setup is really easy.

Reviewer 2: ```suggestion
You can read more about each of the blocks in the [project's GitHub Repository](https://github.com/xmartlabs/android-snapshot-publisher#setup), but I promise you that the setup is really easy.
```

After the setup process is finished, the plugin defines some gradle tasks that can be run.
The naming convention is as follows: `[action][Variant][BuildType]`.

> For example, `publishSnapshotGooglePlayStagingRelease` will be generated if the app has a `staging` flavor and `release` build type.

Reviewer: Maybe make clear that with flavor you mean `Variant`
```suggestion
For example, `publishSnapshotGooglePlayStagingRelease` will be generated if the app has a `staging` variant and `release` build type.
```

To find available tasks, run `./gradlew tasks` and look under the "Snapshot Publishing" section.

The main available tasks are:

- `publishSnapshotFabric[Variant][BuildType]`: it'll publish a snapshot version in Fabric's Beta.
- `publishSnapshotGooglePlay[Variant][BuildType]`: it'll publish a snapshot version in Google Play.

## Conclusion

> This project was designed to release hot builds quickly and easily without wasting your time.

Reviewer 2: ```suggestion
This project was designed to release quick and valuable builds without wasting your time.
```

> One of the most interesting features of it, is that you can integrate it easily with the CI/CD servers.

Reviewer 2: ```suggestion
One of the most interesting features is that you can integrate it easily with a CI/CD server.
```

> You can schedule a build each day or you can subscribe to some GitHub hook, for instance you can perform the release after a merge is made to some specific branch or when a git tag is created.

Reviewer 2: ```suggestion
For instance, you could schedule a build every night or you could subscribe to some GitHub hook and perform the release after a merge is made or a Git tag created.
```

> We're currently using it and it's helping a lot.

Reviewer 2: ```suggestion
We're currently using it and it's helping us a lot.
```

> We are happy because we don't spend time in releases, our coworkers are happy because they can try the newest version of our different applications and the clients are happy because they are receiving build with much more frequency than before.

Reviewer 2: ```suggestion
We are happy because we don't end up spending time in releases, our coworkers are happy because they can try the newest version of our different applications and the clients are happy because they are receiving builds with much more frequency than before.
```

> For all of these and more I recommend you to give it a try and then tell me what do you think!

Reviewer 2: ```suggestion
For all these reasons and more I recommend you give it a try and then tell me what you think!
```

> We are happy because we don't end up spending time in releases, our coworkers are happy because they can try the newest version of our different applications and the clients are happy because they are receiving builds with much more frequency than before.

Reviewer: ```suggestion
We are happy because we don't end up spending time with releases, our coworkers are happy because they can try the latest version of our different apps and the clients are happy because they are receiving builds more frequently than before. 😄 
```

As a side comment, you could consider adding more emojis to the blog

> For all these reasons and more I recommend you give it a try and then tell me what you think!

Reviewer: ```suggestion
For all these reasons and more, I recommend you give it a try and then tell me what you think!
```

---
layout: post
title: Introducing Android Snapshot Plugin
date: 2019-03-20 12:00:00
author: Matías Irland
categories: Android, Gradle Plugin, Snapshot, Google Play, Fabric Beta
author_id: mirland

---

> We're happy to announce the release of our first [open source **Gradle Android Plugin**](https://github.com/xmartlabs/android-snapshot-publisher), a plugin to **create Android Snapshot versions** in the simplest way we know!

Author: the real long name is "Android Snapshot Publisher", @bryant1410 shouldn't I use it instead of `**Snapshot**`? Maybe it's too long, that do you think?

> We're happy to announce the release of our first [open source **Gradle Android Plugin**](https://github.com/xmartlabs/android-snapshot-publisher), a plugin to **create Android Snapshot versions** in the simplest way we know!

Reviewer 4: Use it, but also reflect it on the repo readme.

> We're happy to announce the release of our first [open source **Gradle Android Plugin**](https://github.com/xmartlabs/android-snapshot-publisher), a plugin to **create Android Snapshot versions** in the simplest way we know!

Author: ```suggestion
We're happy to announce the release of [**Android Snapshot Publisher**, our first open source Android Gradle Plugin](https://github.com/xmartlabs/android-snapshot-publisher), a plugin to **create Android Snapshot versions** in the simplest way we know!
```

Here at [Xmartlabs](https://www.xmartlabs.com/) we love engineering new products, adding new features to existing ones and also improving and fixing up our client's apps.
Our awesome clients are oftentimes so exited about the work we do that they want to try it ASAP.
Furthermore, our QA department or some other team members could also be waiting for our changes to deploy in order to test themselves.
That's why we try to prepare and deliver new *Snapshot Builds* as soon as possible.

Do you find this familiar? If you do, you probably know that preparing and distributing new builds is not too hard but can take up valuable time.
Nowadays the CI/CD servers are of great help as they can compile and deliver our applications with ease.
Additionally, they let you schedule tasks so you don't have to care about tests or linter checks running, they will simply be executed before every deployment.

However, there's always some work that must be done before releasing a new build.
In my opinion two of the most important tasks we have to perform before deploying all builds are:
1. Adding information to track possible issues.

> For example if a crash happened you'll want to know which was the specific build that caused that crash.

Reviewer 3: "If a crash happens..." instead of `if .... happened`

> For example if a crash happened you'll want to know which was the specific build that caused that crash.

Author: ```suggestion
For example, if a crash happens, you'll want to know which was the specific build that caused that crash.
```

> 1. Generating good release notes (including a changelog) to check what changed when compared to previous builds.

Reviewer 3: "....previous (app) version/s"

> 1. Generating good release notes (including a changelog) to check what changed when compared to previous builds.

Author: ```suggestion
1. Generating reasonable release notes (including a changelog) to be aware of what changed with respect to previous versions.
```

This is essential as you must communicate what should be tested.

Based on that we decided to create our very own [**Gradle Android Snapshot Publisher plugin!**](https://github.com/xmartlabs/android-snapshot-publisher)

> ## Which are the main features of this plugin?

Reviewer 3: what's are the plugin's main features?

> ## Which are the main features of this plugin?

Author: ```suggestion
## what're the plugin's main features?
```

It was designed to **prepare and distribute Android Snapshot builds**.

The main features of the preparation process are:
- Update the Android Version Name to keep track of the distributed versions.
The default behavior appends the commit identifier to the Android Version name. For instance, a version name may look similar to `1.0.2-b799fbc`.

> - Create and customize rich release notes based on Git's history.

Reviewer 3: I would suggest to remove default behaviour or implementation details when enumerating benefits or features.  `The default behavior appends the commit identifier to the Android Version name. For instance, a version name may look similar to `1.0.2-b799fbc`.` could be moved to another part of the readme.

> - Create and customize rich release notes based on Git's history.

Author: I think that if we use just ` Update the Android Version Name to keep track of the distributed versions.` it's not clear at all. I think it's not clear what the plugin does. I think we should keep the whole phrase. 
What do you think @mtnbarreto ?

After this preparation work, the plugin will proceed to compile and publish the snapshot build.

> Currently the available sources to deploy the builds are:

Reviewer 3: How important is to split up these 3 steps into preparation and after preparation work? There are just 3 steps involved and are clear enough.

- [Google Play](https://play.google.com/apps/publish)
- [Fabric Beta](https://docs.fabric.io/android/beta/overview.html)

The next picture shows an example of a Build published by the Plugin.
***MOCK AN IMAGE***

## What is the preparation work based on?

The plugin makes good use of your project's Git repository to perform the prep work.
To get the most out of this plugin, you have to be as neat as you can when managing it.

> In my team, we are following a simple but nice Git flow.

Reviewer 3: `In my team` -> `At Xmartlabs`

> In my team, we are following a simple but nice Git flow.

Reviewer 3: Default values for app version and release notes are automatically taken from repository git history.....

> In my team, we are following a simple but nice Git flow.

Author: > In my team -> At Xmartlabs

It's not the Xmartlabs flow, It's the android team flow.
For example I know that some teams are still using merge commits.

GitHub provides 3 predefined ways to do that, "Merge", "Rebase and Merge" or "Squash and Merge" (if you are not using GitHub you can just use Git's commands to get the same result).
Based on how the branch was created it could contain multiple commits but all of them related to only one feature or fix. 
So the main value of that branch is the feature or fix you are introducing.
For this reason, some time ago we decided to start using the "Squash and Merge" option, together with a good and descriptive message.
The key to this flow is that if you check the Git history of your main branches, each commit has an understandable and important value.
This was just an introduction to our Git flow, following it is not mandatory to use this plugin.
However, if you use a similar flow where each commit is of value and has a descriptive message your **Snapshot Build** will be great! 

## How to get started?

The first step is to [include the plugin dependency in your project and apply it in your Android module](https://github.com/xmartlabs/android-snapshot-publisher#installation). It's the typical setup of any plugin.

> Then, the plugin defines a `snapshotPublisher` block where you can add the different setups alongside the Android modules.

Reviewer 3: The first step is to -> First

```groovy
snapshotPublisher {
    version {
        // Version customization
    }
    releaseNotes {
        // Release notes customization
    }
    fabric {
        // Fabric Beta setup
    }
    googlePlay {
        // Google Play setup
    }
}
```

`version`: Defines the Android Version Name customizations for the delivered build.
The default value is the current version name and the short-hash commit, joined by a hyphen.
`releaseNotes`: Defines the release notes customizations.
`fabric`: Defines the configuration needed to deploy the artifacts in Fabric's Beta.
`googlePlay`: Defines the configuration needed to deploy the artifacts in Google Play.

You can read more about each of the blocks in the [project's GitHub Repository](https://github.com/xmartlabs/android-snapshot-publisher#setup), but I promise you that the setup is really easy.

> After the setup process is finished, the plugin defines some gradle tasks that can be run.

Reviewer 4: ```suggestion
After the setup process is finished, the plugin defines some Gradle tasks that can be run.
```

The naming convention is as follows: `[action][Variant][BuildType]`.
For example, `publishSnapshotGooglePlayStagingRelease` will be generated if the app has a `staging` variant and `release` build type.

To find available tasks, run `./gradlew tasks` and look under the "Snapshot Publishing" section.

The main available tasks are:

- `publishSnapshotFabric[Variant][BuildType]`: it'll publish a snapshot version in Fabric's Beta.
- `publishSnapshotGooglePlay[Variant][BuildType]`: it'll publish a snapshot version in Google Play.

## Conclusion

This project was designed to release quick and valuable builds without wasting your time.
One of the most interesting features is that you can integrate it easily with a CI/CD server.
For instance, you could schedule a build every night or you could subscribe to some GitHub hook and perform the release after a merge is made or a Git tag created.
We're currently using it and it's helping us a lot.

> We are happy because we don't end up spending time with releases, our coworkers are happy because they can try the latest version of our different apps and the clients are happy because they are receiving builds more frequently than before. 😄 

Reviewer 4: ```suggestion
We are glad that we don't end up spending significant time with releases, our coworkers are happy because they can try the latest version of our different apps and the clients are pleased to receive builds more frequently than ever before. It's a win-win-win! 😄 
```

Reviewer 3: This project was designed to release quick and valuable builds without wasting your time.

`This project was designed to quickly release app builds by doing it automatically according the git history.'
> You will never again have to waste your time checking what is the previous app build or which was the features and fixes introduced in the new build.  (or something like that)

Author: > `This project was designed to quickly release app builds by doing it automatically according the git history.'
So so, it has not to be automatically, maybe you run the command manually. The idea is that you don't waste time when you are releasing a new build

> For all these reasons and more, I recommend you give it a try and then tell me what you think!

Reviewer 4: ```suggestion
For all these reasons and more, I recommend you give it a try and then tell us what are your thoughts!
```

Reviewer 3: Try to use we instead of I.

> For all these reasons and more, I recommend you give it a try and then tell me what you think!

Author: ```suggestion
For all these reasons and more, we recommend you give it a try and then tell us what are your thoughts!
```
